### ISSUE TITLE
Recursive MDReplicatedMembers

### ISSUE THREAD

**Issue by Beider** [DoubleDeez/MDFramework#65]
## The Problem
Issue #62 introduced data converters to the MDReplicatedMember, it also introduced a generic data converter that can convert any type of class.

However currently recursion is not supported, so it is impossible to have a class with another replicated class inside. It is also impossible to have a MDList with another MDList inside it. Nor can you have a MDList inside a replicated class.

## The Solution
This will require some more refactoring and wider changes to the system. 
Currently most network methods use a `params object[] Parameters` to send data across the network. The MDList and the data converters send their data as index + value pairs.

```
[0] = <index>
[1] = <data>
[2] = <index>
etc...
```

The most logical way to fix this would be to modify the sending so the index part also contains length of the data. So something like,
```
[0] = <index>#<length>
[1] = <data>
[2] = Data
etc...
```

In addition we also need to make member registration recursive and more generic. MDList, class data converter should support further converters / MDList inside them. For this we may need to implement some kind of new interface or data structure. We should also take care so we only send the data that is changed.

Eg. If we got something like this `MDList<MDList<CustomClass>>`, when a value in a custom class in the inner list is changed only the change for that value should be sent. That means the message may look something like this

```
[0] <index for outer MDList>#3
[1] <index for inner MDList>#2 // This would be the data from previous line
[2] <index for CustomClass member that changed>#1
[3] <value for the CustomClass member>
```

**Comment by Beider** [comment]
In the scope of this bug please add a storage for converters supplied by MDReplicatedMemberSetting to MDStatics.

This should probably be something like,
`Dictionary<ConverterTypeIncludingGenericType, ConverterInstance>`

Then make both MDReplicatedMember and MDList use this for lookup before instancing. Probably move the code to MDStatics.

**Comment by Beider** [comment]
So to outline this solution further, "all" that needs to be done here is to adjust the default IMDDataConverter for custom classes. It already does partial replication of only changed things and has a ShouldBeReplicated function

- Make it resolve the correct data converter for all members
- Make it send all data as described above, where we got command#length of data->data->etc..
- Once this works ensure that MDList can also be a child of a custom class, with another custom class inside it that contains MDList etc...
- Partial replication should work fine so we only need to send what has changed.
- Once this is solved it will pretty much also solve #76, all that needs to be done for that one is then make sure we always send the complete item. And also that we don't interfer with something like the MDList so we don't consume commands waiting to be processed.
